### ISSUE TITLE
Boost from average.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#4321]
This is a discussion for implementing `BoostFromAverage` from LightGBM.  There are two use cases for this:
- Some objectives don't have Hessian,  which results in suboptimal update path.
- When design matrix is empty, we still output something, related to #4120.

We already have `base_margin` which can be used to implement this feature.  I'm not sure about whether this is indeed useful in XGBoost depending on whether we want to deal with above cases.  @RAMitchell we discussed this before and discarded the idea back then.  Let's see how the others think about this. ;-)

**Comment by trivialfis** [comment]
ping @hcho3 @CodingCat .

**Comment by RAMitchell** [comment]
I think there is a strong case for boosting from the average. As an example consider a linear regression problem where the mean of the labels is 1e10 but the variance is 1e-5. Xgboost's training with any learning rate less than 1.0 will be dominated by trying to reach the mean.

Note that 'average' has different meaning for each objective function, LightGBM implements this as a part of their objective function interface.

**Comment by trivialfis** [comment]
After reading the source code from scikit-learn and lgb (who adopted this method from scikit-learn I think), so the general scheme is to construct trees solely by first order gradient, then set leaves value based on objective specific "average" discarding previous training result, for MAE this is 0.5 quantile.

Now I remember why we @RAMitchell didn't implement this, primarily because we don't know how to pass those data without creating too much weirdly connected interfaces.
